### PR TITLE
Authentication support based on username and password

### PR DESCRIPTION
**Pull request by shanjunmei**
Hi DR,I tried to make some modifications to make it support username and password authentication, and added a compiled API based on http, but I am not familiar with the ssh protocol itself,so need your help to improve it

### CODE REVIEW

**Comment by NHAS** [comment]
Hi this pull request makes a large number of changes, a lot of which remove key sanity checks from the client and server code. 
It also does not meet the stated goal of #78 to have connections *to* the clients be password protected.

As it stands this commit would break the embedding of authentication material in the `--link` command, the gathering of host and username information from the client device and force clients to use a username and password to connect to the server. 

This pull request needs significant work to meet the goal stated in #78, and to reverse the changes to project that would cause instability and reduction of the currently offered features.

As such this will not be merged until this work has been done. 

A final note:

Asymmetric encryption is far more suitable as a authentication mechanism than username and passwords. Embedding a username and password into client binaries does not better identify them, and it reduces the overall security of both the server and client. 

Removing, breaking, or otherwise changing the public key authentication requirement interferes with a core part of this project.

File: internal/client/client.go [hunk 1]
```diff
@@ -100,30 +105,33 @@ func Run(addr, fingerprint, proxyAddr string) {
 
 	sshPriv, sysinfoError := keys.GetPrivateKey()
 	if sysinfoError != nil {
-		log.Fatal("Getting private key failed: ", sysinfoError)
+		log.Println("Getting private key failed: ", sysinfoError)
```
Comment (NHAS) [inline internal/client/client.go:108]: Clients still need to generate a private key, and if they cant that is a fatal error.

File: internal/client/client.go [hunk 2]
```diff
@@ -100,30 +105,33 @@ func Run(addr, fingerprint, proxyAddr string) {
...
+			username = userInfo.Username
+		}
 	}
 
-	hostname, sysinfoError := os.Hostname()
-	if sysinfoError != nil {
-		hostname = "Unknown Hostname"
-		l.Warning("Couldnt get host name: %s", sysinfoError)
-	}
+	/*
```
Comment (NHAS) [inline internal/client/client.go:122]: This removes core functionality from RSSH, as it will no longer be able to show what hostname the device has.

File: internal/client/client.go [hunk 3]
```diff
@@ -100,30 +105,33 @@ func Run(addr, fingerprint, proxyAddr string) {
...
 	l := logger.NewLog("client")
 
-	var username string
-	userInfo, sysinfoError := user.Current()
-	if sysinfoError != nil {
-		l.Warning("Couldnt get username: %s", sysinfoError.Error())
-		username = "Unknown"
-	} else {
-		username = userInfo.Username
+	if username == "" {
```
Comment (NHAS) [inline internal/client/client.go:113]: This is confusing, as it says "if I have a username specified then use the specified user name otherwise get the system username" which makes very little sense, and again hides the real world user that the RSSH client starts in, which reduces the featureset of RSSH

File: internal/client/client.go [hunk 4]
```diff
@@ -100,30 +105,33 @@ func Run(addr, fingerprint, proxyAddr string) {
...
+		hostname, sysinfoError := os.Hostname()
+		if sysinfoError != nil {
+			hostname = "Unknown Hostname"
+			l.Warning("Couldnt get host name: %s", sysinfoError)
+		}
+	*/
 
 	config := &ssh.ClientConfig{
-		User: fmt.Sprintf("%s.%s", username, hostname),
+		User: username, //fmt.Sprintf("%s.%s", username, hostname),
```
Comment (NHAS) [inline internal/client/client.go:131]: Omits important information

File: internal/client/client.go [hunk 5]
```diff
@@ -100,30 +105,33 @@ func Run(addr, fingerprint, proxyAddr string) {
...
 		Auth: []ssh.AuthMethod{
 			ssh.PublicKeys(sshPriv),
```
Comment (NHAS) [inline internal/client/client.go:133]: This will break now that the check ssh key gen failure check is gone

File: internal/client/client.go [hunk 6]
```diff
@@ -100,30 +105,33 @@ func Run(addr, fingerprint, proxyAddr string) {
...
+			ssh.Password(password),
```
Comment (NHAS) [inline internal/client/client.go:134]: AS discussed in #79, the goal isnt to have clients connect to the server with a username and password. As this offers no benefit to the project, and is a weaker authentication mechanism than public/private keypair. This will be rejected if used.

File: internal/config/config.go [hunk 7]
```diff
@@ -0,0 +1,26 @@
+package config
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"log"
+)
+
+type Config struct {
```
Comment (NHAS) [inline internal/config/config.go:9]: What is this for? Add comments

File: internal/server/server.go [hunk 8]
```diff
@@ -66,7 +66,7 @@ func Run(addr, dataDir, connectBackAddress string, insecure, enabledWebserver bo
 
 	private, err := CreateOrLoadServerKeys(privateKeyPath)
 	if err != nil {
-		log.Fatal(err)
+		log.Println("Create or load server key fail :", err)
```
Comment (NHAS) [inline internal/server/server.go:69]: This will break later things that expect the server to have a private key

File: internal/server/sshd.go [hunk 9]
```diff
@@ -144,21 +145,19 @@ func StartSSHServer(sshListener net.Listener, privateKey ssh.Signer, insecure bo
 	log.Printf("Loading authorized keys from: %s\n", authorizedKeysPath)
 	authorizedControllers, err := readPubKeys(authorizedKeysPath)
 	if err != nil {
-		log.Fatal(err)
+		log.Println("read authorizekeys fail :", err)
```
Comment (NHAS) [inline internal/server/sshd.go:148]: This will break things that expect authorizedControllers to be non-nil/non-empty

File: internal/server/sshd.go [hunk 10]
```diff
@@ -144,21 +145,19 @@ func StartSSHServer(sshListener net.Listener, privateKey ssh.Signer, insecure bo
...
 	}
 
 	if _, err := os.Stat("downloads"); err != nil && os.IsNotExist(err) {
 		os.Mkdir("downloads", 0700)
 		log.Println("Created downloads directory")
 	}
 
+	conf := config.LoadConfig(filepath.Join(dataDir, "config.json"))
```
Comment (NHAS) [inline internal/server/sshd.go:156]: What does this do?

File: internal/server/sshd.go [hunk 11]
```diff
@@ -144,21 +145,19 @@ func StartSSHServer(sshListener net.Listener, privateKey ssh.Signer, insecure bo
...
+
 	clients, err := readPubKeys(authorizedControlleeKeysPath)
 	if err != nil {
-		if !insecure {
-			log.Fatal(err)
-		} else {
-			log.Println(err)
-		}
+		log.Println("read authorizedControlleeKey fail :", err)
```
Comment (NHAS) [inline internal/server/sshd.go:160]: This will break things that expect authorized controllee keys to be non-empty. Also, we're not removing public key auth from this project. 
This will also mean it is impossible to identify if a client is a controllable device, or a controller.

File: internal/server/sshd.go [hunk 12]
```diff
@@ -231,6 +230,20 @@ func StartSSHServer(sshListener net.Listener, privateKey ssh.Signer, insecure bo
 			}, nil
 
 		},
+		PasswordCallback: func(conn ssh.ConnMetadata, password []byte) (*ssh.Permissions, error) {
```
Comment (NHAS) [inline internal/server/sshd.go:233]: We're not adding password based authentication to the RSSH server.

File: internal/server/webhooks/server.go [hunk 13]
```diff
@@ -36,7 +36,8 @@ func StartWebhooks(config string) {
 
 	err = json.Unmarshal(activeWebhooks, &recipients)
 	if err != nil {
-		log.Fatal(err)
+		log.Println(err)
+		return
```
Comment (NHAS) [inline internal/server/webhooks/server.go:40]: Why?

File: internal/server/webserver/api.go [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/NHAS/reverse_ssh/internal/server/webserver/shellscripts"
+)
+
+func download(w http.ResponseWriter, req *http.Request) {
```
Comment (NHAS) [inline internal/server/webserver/api.go:14]: If you're going to split out the build and download functionalities, please do this in another pull request. This pull request is to resolve/add #79, not do this

File: internal/server/webserver/buildmanager.go [hunk 15]
```diff
@@ -154,7 +154,7 @@ func Build(goos, goarch, suppliedConnectBackAdress, fingerprint, name string, sh
 		return "", err
 	}
 
-	buildArguments = append(buildArguments, fmt.Sprintf("-ldflags=-s -w -X main.destination=%s -X main.fingerprint=%s -X github.com/NHAS/reverse_ssh/internal.Version=%s", suppliedConnectBackAdress, fingerprint, f.Version))
+	buildArguments = append(buildArguments, fmt.Sprintf("-ldflags=-s -w -X main.destination=%s -X main.fingerprint=%s -X github.com/NHAS/reverse_ssh/internal.Version=%s -X github.com/NHAS/reverse_ssh/internal.server.webserver.username=%s -X github.com/NHAS/reverse_ssh/internal.server.webserver.password=%s", suppliedConnectBackAdress, fingerprint, f.Version, username, password))
```
Comment (NHAS) [inline internal/server/webserver/buildmanager.go:157]: The additional link-time insertions will not be added to the client binary as they are targeting the internal web server? 
I dont think this will work

**Review by NHAS — CHANGES_REQUESTED** [review]

File: internal/server/sshd.go [hunk 16]
```diff
@@ -231,6 +230,20 @@ func StartSSHServer(sshListener net.Listener, privateKey ssh.Signer, insecure bo
...
+			params := make(map[string]string, 0)
+			res, err := Post(conf.ExternalAuthApi, params)
+			if err != nil {
+				return nil, fmt.Errorf("call external auth api %v get error", conf.ExternalAuthApi)
```
Comment (lachlan2k) [inline internal/server/sshd.go:237]: Does this even do anything?? `params` is just an empty map? And if the HTTP request doesn't give an error, it just allows auth...?

Comment (NHAS) [reply internal/server/sshd.go:237]: Ah good point, this would not working as an authentication method, I was a little bit focused on the bits that were removing public key authentication which is explicitly against what I want to do.

Comment (shanjunmei) [reply internal/server/sshd.go:239]: Sorry, this was an oversight, forgot to wear the username and password, I already have a new commit, but not sure how to update the pr, I'm still working on it

File: internal/server/webserver/buildmanager.go (hunk 15 above)
Comment (shanjunmei) [reply internal/server/webserver/buildmanager.go:157]: This is to support each client to embed an independent username and password. I have created an independent web api, which can be called by the user passing in the username and password.

File: internal/server/webserver/api.go (hunk 14 above)
Comment (shanjunmei) [reply internal/server/webserver/api.go:14]: Because before this, there is only one downloaded api, and there is no need for an independent compilation api

File: internal/server/webhooks/server.go (hunk 13 above)
Comment (shanjunmei) [reply internal/server/webhooks/server.go:40]: webhook is not a critical feature, it should allow failure

File: internal/server/sshd.go (hunk 12 above)
Comment (shanjunmei) [reply internal/server/sshd.go:233]: #79 but i think you already agree on this..

File: internal/server/sshd.go (hunk 11 above)
Comment (shanjunmei) [reply internal/server/sshd.go:160]: You can still verify based on the key, but this pr has added a password-based verification method, whether it is the control end or the controlled end, we can get it in the external api

File: internal/client/client.go (hunk 5 above)
Comment (shanjunmei) [reply internal/client/client.go:133]: Consider handling this malformed username on the client side if it affects public key generation

File: internal/client/client.go (hunk 4 above)
Comment (shanjunmei) [reply internal/client/client.go:131]: Which user started it? In fact, once we are connected, there are many commands that can be obtained? So actually it's not that important here?

File: internal/client/client.go (hunk 3 above)
Comment (shanjunmei) [reply internal/client/client.go:113]: Really start the user, once we are connected, we can still get a lot of commands, here is to support, compile an independent user name and password for each client

File: internal/client/client.go (hunk 2 above)
Comment (shanjunmei) [reply internal/client/client.go:122]: But this host name is not really important? Because this usage scenario is deployed in an unfamiliar subnet, but we can consider dealing with the authentication user name on the client side

File: internal/client/client.go (hunk 1 above)
Comment (shanjunmei) [reply internal/client/client.go:108]: Once username and password authentication is supported, this key is no longer a key error

**Comment by shanjunmei** [comment]
> 

Regarding the issue of embedding the user name and password into the client, in fact, it will not have much impact on security. Even if someone reverses the client to obtain the user name and password, it can only be used as the controlled end. I believe that no one will be willing to do so.  the control end can still connect to the server through the key

i have a update to fix key authentication

File: internal/server/webhooks/server.go (hunk 13 above)
Comment (NHAS) [reply internal/server/webhooks/server.go:40]: That isnt what this code does. While agree webhooks arent a critical feature, if someone has a configuration file that indicates they *want* webhooks, but we cant parse it, thats an error that is critical and they need to sort it out.

File: internal/server/sshd.go (hunk 12 above)
Comment (NHAS) [reply internal/server/sshd.go:233]: No. I was agreeing that requiring a password before a user can **connect** to a device is a good idea.

File: internal/client/client.go (hunk 2 above)
Comment (NHAS) [reply internal/client/client.go]: That isnt solid reasoning, in doing this you're giving less information to the user who is looking at the client. Plus changes have not been made on the server side to accommodate this.

File: internal/server/webhooks/server.go (hunk 13 above)
Comment (shanjunmei) [reply internal/server/webhooks/server.go:40]: okie ,fixed it ,just wait a new pr

File: internal/client/client.go (hunk 3 above)
Comment (NHAS) [reply internal/client/client.go:113]: I am not keen on having the clients have their own username and password to authenticate to the server. It adds no benefit over public key authentication and is in-fact worse.

File: internal/client/client.go (hunk 4 above)
Comment (NHAS) [reply internal/client/client.go]: It omits the hostname, and the real user that the rssh binary is running as

File: internal/client/client.go (hunk 5 above)
Comment (NHAS) [reply internal/client/client.go:131]: This is incorrect, the username is given to the server. Handling the username malformation doesnt even come into the equation here?

File: internal/server/sshd.go (hunk 11 above)
Comment (NHAS) [reply internal/server/sshd.go:160]: No. You pull request removes all ability to verify a client binary by public key. Im not going to accept work that then requires work on my behalf to add. 
Additionally, as I've said in other comments, this project is not moving over to using password authentication for the client devices to connect to the server.

File: internal/client/client.go (hunk 3 above)
Comment (shanjunmei) [reply internal/client/client.go:113]: if you no need this feature ,just keep build without pass username and password,nothing will be changed

File: internal/server/webserver/buildmanager.go (hunk 15 above)
Comment (NHAS) [reply internal/server/webserver/buildmanager.go:157]: So, effectively if anyone extracts the username and password from any client binary they'll be able to connect to a hosted RSSH server? 

Also, just reiterating, I am not moving to letting clients do password based auth to the server.

File: internal/client/client.go (hunk 2 above)
Comment (lachlan2k) [reply internal/client/client.go]: > But this host name is not really important? Because this usage scenario is deployed in an unfamiliar subnet, but we can consider dealing with the authentication user name on the client side

Identifying the hostname of a client is one of the most cruicial parts of many RSSH usecases, including mine.

Please don't remove features from open-source projects just because they don't meet *your* needs, at the expense of everyone else.

Ontop of that, it is inappropriate to add breaking changes like this in a PR, when it is not related to the functionality in the PR.

PRs should only include the functionality discussed in related issues, and anything else should be clearly disclosed in the body of the PR, with justification.

File: internal/server/webserver/buildmanager.go (hunk 15 above)
Comment (shanjunmei) [reply internal/server/webserver/buildmanager.go:157]: Only the controlled end will embed the user name and password. Even if someone reverses the user name and password, it can only be connected as the controlled end. Even if it is reversed, no one should be willing to connect.

**Comment by NHAS** [comment]
It will impact security. As it currently stands in a usual setup each client binary is generated with a new public/private key pair. 
Thus, is a single client binary is decompiled and the private key extracted it is easy to deny access to that one key. 
However in your solution you would have to invalidate an entire fleet of RSSH clients if the "master" password is leaked.

File: internal/client/client.go (hunk 2 above)
Comment (shanjunmei) [reply internal/client/client.go]: Sorry, I've reverted this in a new commit

**Comment by shanjunmei** [comment]
> It will impact security. As it currently stands in a usual setup each client binary is generated with a new public/private key pair. Thus, is a single client binary is decompiled and the private key extracted it is easy to deny access to that one key. However in your solution you would have to invalidate an entire fleet of RSSH clients if the "master" password is leaked.

There is no such thing as the failure of the master password you mentioned, because the username and password will not be embedded in the server. The server uses the username and password sent by the client to call another API for authentication and obtain the connection. Is it the identification of the server or the controlled end?

**Comment by NHAS** [comment]
So are your client binaries not going to contain the username and password they use to connect to the server?

**Comment by shanjunmei** [comment]
> So are your client binaries not going to contain the username and password they use to connect to the server?

Unless the control terminal itself connects to the server through this client, and he actively exposes his own client with control terminal permissions, and even if this problem occurs, we can still make it downgrade by adjusting the permissions in the external api the controlled side, or even delete it

**Comment by NHAS** [comment]
Right, I think I vaguely get what you're saying which is:
Enable the RSSH server to ask an API if a client/user should be connected

Yes?

To that end, using unique public keys per user is still easier than using a username + password combo. As your api would do the following: 

"Oh I got this public key, I'll look it up in the API, API says this is a client binary and is allowed, cool, I'll let it connect"

Or 

"The API says this public key is a user, I'll let them connect to the RSSH server instead"

Or 

"The API does recognise this public key, it is denied"

File: internal/server/sshd.go (hunk 16 above)
Comment (lachlan2k) [reply internal/server/sshd.go:244]: What is your intention for the response to the POST request?

You just take the `res`, and assign that as the `type` in the permissions extensions? If the API returns "client", it will allow a client binary to connect with SSH. If it returns "user", it will allow a human operator to connect. 

How would the server respond to an authentication failure? I don't see any handling for that? If the server responds with something other than "client" or "user", then even though `acceptConn` would close the connection, the user is still allowed to authenticate with completely incorrect credentials...

Frankly, I find it quite concerning that the first version of this code didn't even send the credentials, and the second version doesn't even handle the response properly, allowing ANY username or password to authenticate...

Comment (shanjunmei) [reply internal/server/sshd.go:244]: Only the correct user can pass the authorization and return the clientType, otherwise it will return non-200, and after the request is responded to, all non-200 will return an error

**Comment by shanjunmei** [comment]
> Right, I think I vaguely get what you're saying which is: Enable the RSSH server to ask an API if a client/user should be allowed connected.
> 
> Yes?
> 
> To that end, using unique public keys per user is still easier than using a username + password combo. As your api would do the following:
> 
> "Oh I got this public key, I'll look it up in the API, API says this is a client binary and is allowed, cool, I'll let it connect"
> 
> Or
> 
> "The API says this public key is a user, I'll let them connect to the RSSH server instead"
> 
> Or
> 
> "The API does recognise this public key, it is denied"

Yes, your understanding is correct, it's just that the key generation is relatively troublesome, anyway, if you are not happy with it, I just close this pr

File: internal/server/sshd.go (hunk 16 above)
Comment (lachlan2k) [reply internal/server/sshd.go:244]: > all non-200 will return an error

This is the opposite of the truth

> A non-2xx status code doesn't cause an error. 

https://pkg.go.dev/net/http#Client.Do

Comment (shanjunmei) [reply internal/server/sshd.go:244]: i mean we can implement it in api itself,any way i have a new commit is meet your wanna
